examples/touch/main.go

```diff
@@ -0,0 +1,253 @@
+// Copyright 2021 The Ebiten Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build example
+
+package main
+
+import (
+	"bytes"
+	"image"
+	_ "image/jpeg"
+	"log"
+	"math"
+
+	"github.com/hajimehoshi/ebiten/v2"
+	"github.com/hajimehoshi/ebiten/v2/ebitenutil"
+	"github.com/hajimehoshi/ebiten/v2/examples/resources/images"
+	"github.com/hajimehoshi/ebiten/v2/inpututil"
+)
+
+// distance between points a and b.
+func distance(xa, ya, xb, yb int) float64 {
+	x := math.Abs(float64(xa - xb))
+	y := math.Abs(float64(ya - yb))
+	return math.Sqrt(x*x + y*y)
+}
+
+const (
+	screenWidth  = 640
+	screenHeight = 480
+)
+
+var (
+	gophersImage *ebiten.Image
+)
+
+type touch struct {
+	originX, originY int
+	currX, currY     int
+	duration         int
+	wasPinch, isPan  bool
+}
+
+type pinch struct {
+	id1, id2 ebiten.TouchID
+	originH  float64
+	prevH    float64
+}
+
+type pan struct {
+	id ebiten.TouchID
+
+	prevX, prevY     int
+	originX, originY int
+}
+
+type tap struct {
+	X, Y int
+}
+
+type Game struct {
+	x, y float64
+	zoom float64
+
+	touches map[ebiten.TouchID]*touch
+	pinch   *pinch
+	pan     *pan
+	taps    []tap
+}
+
+func (g *Game) Update() error {
+	g.taps = []tap{}
```

Comment:
This initialization is not needed (appending works with nil)

Reply:
The any taps from previous frames need to be removed here, but we can definitely assign `nil` instead of an empty slice.

Reply:
Ah, you're right. This should be initialized every tick.

Then, I'd use `g.taps = g.taps[:0]` to avoid allocations when appending. This should work even if `g.taps` is nil.